### PR TITLE
Implement NoProof for json tests and update tests reference (replaces #9744)

### LINKED ISSUE #9590: Currently failing JSON tests
** Note **

Currently this issue refers to an unmerged PR #9570 (early creation to allow reference from the exception json). It should only be considered active after #9570.

** end Note **

Umbrella issue to follow current json-test issues.

All those failing test are currently **skipped** by the CI.

When one test is fixed (either by an update of ethereum/test submodule or an actual parity fix), the corresponding skipped exception should be removed from the [file](https://github.com/paritytech/parity-ethereum/ethcore/res/ethereum/test-issues/currents.json) configuring those exceptions.

Current tags to qualify those issues are:

- `ToInvestigate` : no investigation done yet.
- `IssueWaiting` : the issue is likely to be solve in future ethereum/test generations. 

An indicative number of failure is also added. 

**Last synch**

https://github.com/ethereum/tests/commit/3f5febc901913ef698f1b09dda8705babd729e4a

**Block issues**

- [ ] stCreateTest : `ToInvestigate` : 1
- [ ] stCreate2 : `IssueWaiting` : 151 (all), the issue seems to be present for other client to (see ethereum/tests gitter).
- [ ] bcStateTest: `ToInvestigate` : 10  Note that those are not only for constantinople and are related to contract sucide #9440 could be related or a bugfix from #9319
- [ ] stEIP158Specific: `ToInvestigate` : 3
- [ ] stPreCompiledContracts: `ToInvestigate` : 2
- [ ] stReturnDataTest: `ToInvestigate` : 3
- [ ] stReturnDataTest: `ToInvestigate` : 3
- [ ] stSpecialTest: `ToInvestigate` : 1 ('push32withoutByte_d0g0v0_Constantinople')

**State issues**

- [ ] `stCreateTest`: `IssueWaiting` : 1 (passing with https://github.com/ethereum/tests/pull/511 at baf0eb45c1c02fb56612028bef3ce50932779c05)
- [ ] `stCreateTest2`: `ToInvestigate` : 1 (RevertInCreateInInitCreate2 - constantiople 1/1)
- [ ] `stEIP150Specific`: `IssueWaiting` : 1 (passing with https://github.com/ethereum/tests/pull/511 at baf0eb45c1c02fb56612028bef3ce50932779c05)
- [ ] `stInitCodeTest`: `IssueWaiting` : 1 Running test with https://github.com/ethereum/tests/pull/511 at 67f90e893326f9bde359e74492313269708d5e9a seems to fix it
- [ ] `stPreCompiledContracts` : `IssueWaiting` : 80 all modexp test failling. Running test with https://github.com/ethereum/tests/pull/511 at 67f90e893326f9bde359e74492313269708d5e9a seems to fix it
- [ ] `stRevertTest` : `ToInvestigate` : 14 (all constantinople related). (only 1 error ('RevertInCreateInInit') with https://github.com/ethereum/tests/pull/511 at baf0eb45c1c02fb56612028bef3ce50932779c05)
- [ ] `stRevertTest - RevertOpcodeMultipleSubCalls` : `IssueWaiting` >20 (passing with https://github.com/ethereum/tests/pull/511 at baf0eb45c1c02fb56612028bef3ce50932779c05)
- [ ] `stStaticCall` : `IssueWaiting` : 14 (all constantinople related) . Running test with https://github.com/ethereum/tests/pull/511 at 67f90e893326f9bde359e74492313269708d5e9a seems to fix it, but at next commit baf0eb45c1c02fb56612028bef3ce50932779c05 back to 13. Ok at  8a1e249a739aa0f2f3532a679993067b3bbb8f3c.
- [ ] `stZeroKnowledge` : `IssueWaiting` : all https://github.com/ethereum/tests/issues/512

### PR DESCRIPTION
**Pull request by cheme**
Closes #9590 .

Update to latest ethereum/tests .
Fix cost for constantinople test json.
Implements 'NoProof' engine from ethereum/tests#464 .

At this point with this PR everything pass except the case spotted by @holiman this week in https://gist.github.com/holiman/0154f00d5fcec5f89e85894cbb46fcb2 . 

I skip those tests. For completeness, @sorpaas indicated that it is not an easy fix : so since those case cannot happen on a mainchain I did not open and reference an issue for them (in the json defining tests to skip I only added a comment but not issue reference).

Note for reviewer:
- an other version of this PR exists with previous review comments is #9744 , this PR replaces it . 
- when running NoProof with CanonNoSeal, the upper gas limit verification is required : I switch it from verify_block_basic to the header check. I got  to put another trait method for it, a lighter change would be to make this check for every engine but it would change a bit the existing behavior (might not be an issue).

### CODE REVIEW

File: ethcore/src/ethereum/ethash.rs [hunk 1]
```diff
@@ -222,6 +222,8 @@ impl Engine<EthereumMachine> for Arc<Ethash> {
 
 	fn maximum_uncle_count(&self, _block: BlockNumber) -> usize { 2 }
 
+	fn maximum_gas_limit(&self) -> Option<U256> { Some(0x7fffffffffffffffu64.into()) }
```
Comment (niklasad1) [inline ethcore/src/ethereum/ethash.rs]: Can you use `underscores` for this number to make it a little bit more readable?

File: ethcore/src/verification/verification.rs [hunk 2]
```diff
@@ -281,6 +281,11 @@ pub fn verify_header_params(header: &Header, engine: &EthEngine, is_full: bool,
 	if header.gas_limit() < &min_gas_limit {
 		return Err(From::from(BlockError::InvalidGasLimit(OutOfBounds { min: Some(min_gas_limit), max: None, found: header.gas_limit().clone() })));
 	}
+	if let Some(limit) = engine.maximum_gas_limit() {
+		if header.gas_limit() > &limit {
+			return Err(From::from(::error::BlockError::InvalidGasLimit(OutOfBounds { min: None, max: Some(limit), found: header.gas_limit().clone() })));
```
Comment (niklasad1) [inline ethcore/src/verification/verification.rs]: `clone()` on copy-type 
```suggestion
			return Err(From::from(::error::BlockError::InvalidGasLimit(OutOfBounds { min: None, max: Some(limit), found: *header.gas_limit() })));
```

File: rpc/src/v1/tests/eth.rs [hunk 3]
```diff
@@ -84,7 +85,16 @@ struct EthTester {
...
+
+		let tester = if ::ethjson::blockchain::Engine::NoProof == chain.engine {
+			let mut config = ClientConfig::default();
+			config.verifier_type = VerifierType::CanonNoSeal;
+			config.check_seal = false;
+			Self::from_spec_conf(make_spec(chain), config)
+		} else {
+			Self::from_spec(make_spec(chain))
+		};
+
```
Comment (andresilva) [inline rpc/src/v1/tests/eth.rs]: stray newline

File: rpc/src/v1/tests/eth.rs [hunk 4]
```diff
@@ -101,14 +111,20 @@ impl EthTester {
 	}
 
 	fn from_spec(spec: Spec) -> Self {
+		let config = ClientConfig::default();
+		Self::from_spec_conf(spec, config)
+	}
+
+	fn from_spec_conf(spec: Spec, config: ClientConfig) -> Self {
+
```
Comment (andresilva) [inline rpc/src/v1/tests/eth.rs]: stray newline

File: ethcore/res/ethereum/tests-issues/currents.json [hunk 5]
```diff
@@ -1,468 +1,42 @@
...
-					"subnumbers": ["2","3"],
-					"chain": "Constantinople (test)"
-				},
-				"create2callPrecompiles": {
-					"subnumbers": ["*"],
-					"chain": "Constantinople (test)"
 				}
 			}
 		}
+
```
Comment (andresilva) [inline ethcore/res/ethereum/tests-issues/currents.json]: stray newline. Actually would be nice if you could run a pretty printer on this file :)

**Pull request merged**